src/CdrWriter.ts

```diff
@@ -241,37 +252,46 @@ export class CdrWriter {
     // M_FLAG is the value of the Must Understand option for the member
     const mustUnderstandFlag = mustUnderstand ? 1 << 31 : 0;
     // LC is the value of the Length Code for the member.
-    let lengthCode: number | undefined;
-    switch (objectSize) {
+    const finalLengthCode = lengthCode ?? getLengthCodeForObjectSize(objectSize);
+
+    const header = mustUnderstandFlag | (finalLengthCode << 28) | id;
+
+    this.uint32(header);
+
+    switch (finalLengthCode) {
+      case 0:
       case 1:
-        lengthCode = 0;
-        break;
       case 2:
-        lengthCode = 1;
+      case 3: {
+        const shouldBeSize = lengthCodeToObjectSizes[finalLengthCode];
+        if (objectSize !== shouldBeSize) {
+          throw new Error(
+            `Cannot write a length code ${finalLengthCode} header with an object size not equal to ${shouldBeSize}`,
+          );
+        }
         break;
+      }
+      // When the length code is > 3 the header is 8 bytes because of the NEXTINT value storing the object size
       case 4:
-        lengthCode = 2;
+      case 5:
+        this.uint32(objectSize);
         break;
-      case 8:
-        lengthCode = 3;
+      case 6:
+        if (objectSize % 4 !== 0) {
+          throw new Error(
+            "Cannot write a length code 6 header with an object size that is not a multiple of 4",
+          );
+        }
+        this.uint32(Math.floor(objectSize / 4));
```

Comment:
(assuming this works?)
```suggestion
        this.uint32(objectSize >> 2);
```